lib/src/models/axi4_bfm/axi4_subordinate.dart

```diff
@@ -0,0 +1,654 @@
+        );
+    for (var i = 0; i < channels.length; i++) {
+      if (channels[i].hasRead) {
+        _dataReadResponseMetadataQueue.add([]);
+        _dataReadResponseDataQueue.add([]);
+        _dataReadResponseErrorQueue.add([]);
+        _dataReadResponseIndex.add(0);
+        _readAddrToChannel[i] = _dataReadResponseMetadataQueue.length - 1;
+      }
+      if (channels[i].hasWrite) {
+        _writeMetadataQueue.add([]);
+        _writeReadyToOccur.add(false);
+        _writeAddrToChannel[i] = _writeMetadataQueue.length - 1;
+      }
+    }
+  }
+
+  @override
+  Future<void> run(Phase phase) async {
+    unawaited(super.run(phase));
+
+    sIntf.resetN.negedge.listen((event) {
+      storage.reset();
+      for (var i = 0; i < channels.length; i++) {
+        if (channels[i].hasRead) {
+          _dataReadResponseMetadataQueue[_readAddrToChannel[i]!].clear();
+          _dataReadResponseDataQueue[_readAddrToChannel[i]!].clear();
+          _dataReadResponseErrorQueue[_readAddrToChannel[i]!].clear();
+          _dataReadResponseIndex[_readAddrToChannel[i]!] = 0;
+        }
+        if (channels[i].hasWrite) {
+          _writeMetadataQueue[_writeAddrToChannel[i]!].clear();
+          _writeReadyToOccur[_writeAddrToChannel[i]!] = false;
+        }
+      }
+    });
+
+    // wait for reset to complete
+    await sIntf.resetN.nextPosedge;
+
+    while (!Simulator.simulationHasEnded) {
+      await sIntf.clk.nextNegedge;
+      for (var i = 0; i < channels.length; i++) {
+        if (channels[i].hasRead) {
+          _driveReadReadys(index: i);
+          _respondRead(index: i);
+          _receiveRead(index: i);
+        }
+        if (channels[i].hasWrite) {
+          _driveWriteReadys(index: i);
+          _respondWrite(index: i);
+          _captureWriteData(index: i);
+          _receiveWrite(index: i);
+        }
+      }
+    }
+  }
+
+  /// Calculates a strobed version of data.
+  static LogicValue _strobeData(
+          LogicValue originalData, LogicValue newData, LogicValue strobe) =>
+      [
+        for (var i = 0; i < strobe.width; i++)
+          (strobe[i].toBool() ? newData : originalData)
+              .getRange(i * 8, i * 8 + 8)
+      ].rswizzle();
+
+  // find the index of the region that the provided address falls in
+  // if none, return -1
+  int _checkRegion(LogicValue addr) {
+    for (var j = 0; j < ranges.length; j++) {
+      if ((addr >= ranges[j].start).toBool() &&
+          (addr < ranges[j].end).toBool()) {
+        return j;
+      }
+    }
+    return -1;
+  }
+
+  // assesses the input ready signals and drives them appropriately
+  void _driveReadReadys({int index = 0}) {
+    // for now, assume we can always handle a new request
+    channels[index].rIntf!.arReady.put(true);
+  }
+
+  // assesses the input ready signals and drives them appropriately
+  void _driveWriteReadys({int index = 0}) {
+    // for now, assume we can always handle a new request
+    channels[index].wIntf!.awReady.put(true);
```

Comment:
you probably want to use `inject` for a BFM -- main impact is waveforms will make more sense